beacon-chain/p2p/handshake.go

```diff
@@ -54,37 +56,47 @@ func (s *Service) AddConnectionHandler(reqFunc func(ctx context.Context, id peer
 						// used in production.
 						log.Debug("Not disconnecting peer with unsupported protocol. This may be the DHT node or relay.")
 						s.host.ConnManager().Protect(conn.RemotePeer(), "relay/bootnode")
+						s.peers.SetConnectionState(conn.RemotePeer(), peers.PeerDisconnected)
 						return
 					}
+					s.peers.IncrementBadResponses(conn.RemotePeer())
+					badResponses, err := s.peers.BadResponses(conn.RemotePeer())
+					if err == nil && badResponses == s.peers.MaxBadResponses() {
+						log.Info("Peer has given too many bad responses; will ignore in future")
+					}
+					s.peers.SetConnectionState(conn.RemotePeer(), peers.PeerDisconnecting)
 					if err := s.Disconnect(conn.RemotePeer()); err != nil {
-						log.WithError(err).Errorf("Unable to close peer %s", conn.RemotePeer())
-						return
+						log.WithError(err).Error("Unable to disconnect from peer")
```

Comment:
you need to return from here otherwise the connection state will be `peers.PeerDisconnected` .Also in the event of failure, maybe you will have to change back the connection state to `peers.PeerConnected` .

Reply:
This is another area where the difference between network-level and protocol-level connectivity confuses things.  Although we have failed to disconnect at the network level the peer has failed to successfully handshake so for the protocol level it never reached `Connected`, instead it went `Connecting`->`Disconnecting`->`Disconnected`.

Reply:
if it failed to disconnect, it would still be an eth2 peer , no ? Its still connected at the protocol level

Reply:
Resolving as per conversation.